### PR TITLE
*: reduce tikvrpc resp struct size

### LINKED ISSUE #9641: Redesign huge structure inside tikv RPC go get better formance
There are two huge structure inside tikv RPC module, Allocate a huge structure is quite slow in Go. We should redesign by using interface instead. See https://github.com/pingcap/tidb/blob/master/store/tikv/tikvrpc/tikvrpc.go#L123 and https://github.com/pingcap/tidb/blob/master/store/tikv/tikvrpc/tikvrpc.go#L200

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

ref #9641, try to reduce tikvrpc response struct size

pair with https://github.com/pingcap/tidb/pull/11055

### What is changed and how it works?

same to https://github.com/pingcap/tidb/pull/11055

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - N/A

Side effects

 - N/A

Related changes

 - Need to cherry-pick to the release 3.0

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/11056)

### CODE REVIEW

**Comment by zz-jason** [comment]
would it improve the performance of sysbench?

**Comment by lysu** [comment]
> would it improve the performance of sysbench?

for my laptop  sysbench result is almost same

before

```
[ 5s ] thds: 32 tps: 54502.15 qps: 54502.15 (r/w/o: 54502.15/0.00/0.00) lat (ms,95%): 1.04 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 32 tps: 29278.36 qps: 29278.36 (r/w/o: 29278.36/0.00/0.00) lat (ms,95%): 1.86 err/s: 0.00 reconn/s: 0.00
[ 15s ] thds: 32 tps: 29869.90 qps: 29869.90 (r/w/o: 29869.90/0.00/0.00) lat (ms,95%): 1.82 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 32 tps: 30095.28 qps: 30095.28 (r/w/o: 30095.28/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 32 tps: 30144.97 qps: 30144.97 (r/w/o: 30144.97/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 32 tps: 30264.10 qps: 30264.10 (r/w/o: 30264.10/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 32 tps: 30224.33 qps: 30224.33 (r/w/o: 30224.33/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 32 tps: 30080.71 qps: 30080.71 (r/w/o: 30080.71/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 45s ] thds: 32 tps: 30678.13 qps: 30678.13 (r/w/o: 30678.13/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 32 tps: 30390.39 qps: 30390.39 (r/w/o: 30390.39/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 55s ] thds: 32 tps: 30330.13 qps: 30330.13 (r/w/o: 30330.13/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 32 tps: 30441.17 qps: 30441.17 (r/w/o: 30441.17/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 65s ] thds: 32 tps: 30362.89 qps: 30362.89 (r/w/o: 30362.89/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 32 tps: 30475.42 qps: 30475.42 (r/w/o: 30475.42/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 75s ] thds: 32 tps: 30536.18 qps: 30536.18 (r/w/o: 30536.18/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 80s ] thds: 32 tps: 30131.14 qps: 30131.14 (r/w/o: 30131.14/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 85s ] thds: 32 tps: 30232.38 qps: 30232.38 (r/w/o: 30232.38/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 90s ] thds: 32 tps: 30727.10 qps: 30727.10 (r/w/o: 30727.10/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 95s ] thds: 32 tps: 30407.55 qps: 30407.55 (r/w/o: 30407.55/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 100s ] thds: 32 tps: 30388.25 qps: 30388.25 (r/w/o: 30388.25/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 105s ] thds: 32 tps: 30713.30 qps: 30713.30 (r/w/o: 30713.30/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 110s ] thds: 32 tps: 30004.42 qps: 30004.42 (r/w/o: 30004.42/0.00/0.00) lat (ms,95%): 1.82 err/s: 0.00 reconn/s: 0.00
```

after 

```
[ 5s ] thds: 32 tps: 50844.76 qps: 50844.76 (r/w/o: 50844.76/0.00/0.00) lat (ms,95%): 1.23 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 32 tps: 29793.35 qps: 29793.35 (r/w/o: 29793.35/0.00/0.00) lat (ms,95%): 1.82 err/s: 0.00 reconn/s: 0.00
[ 15s ] thds: 32 tps: 30263.36 qps: 30263.36 (r/w/o: 30263.36/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 32 tps: 30300.04 qps: 30300.04 (r/w/o: 30300.04/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 32 tps: 30094.84 qps: 30094.84 (r/w/o: 30094.84/0.00/0.00) lat (ms,95%): 1.82 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 32 tps: 30832.96 qps: 30832.96 (r/w/o: 30832.96/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 32 tps: 30455.41 qps: 30455.41 (r/w/o: 30455.41/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 32 tps: 30717.74 qps: 30717.74 (r/w/o: 30717.74/0.00/0.00) lat (ms,95%): 1.73 err/s: 0.00 reconn/s: 0.00
[ 45s ] thds: 32 tps: 30882.89 qps: 30882.89 (r/w/o: 30882.89/0.00/0.00) lat (ms,95%): 1.73 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 32 tps: 30913.59 qps: 30913.59 (r/w/o: 30913.59/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 55s ] thds: 32 tps: 30562.89 qps: 30562.89 (r/w/o: 30562.89/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 32 tps: 31180.21 qps: 31180.21 (r/w/o: 31180.21/0.00/0.00) lat (ms,95%): 1.73 err/s: 0.00 reconn/s: 0.00
[ 65s ] thds: 32 tps: 30745.51 qps: 30745.51 (r/w/o: 30745.51/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 32 tps: 30544.59 qps: 30544.59 (r/w/o: 30544.59/0.00/0.00) lat (ms,95%): 1.76 err/s: 0.00 reconn/s: 0.00
[ 75s ] thds: 32 tps: 30080.87 qps: 30080.87 (r/w/o: 30080.87/0.00/0.00) lat (ms,95%): 1.82 err/s: 0.00 reconn/s: 0.00
[ 80s ] thds: 32 tps: 30720.05 qps: 30720.05 (r/w/o: 30720.05/0.00/0.00) lat (ms,95%): 1.79 err/s: 0.00 reconn/s: 0.00
```

in pprof, can not make this part be bottleneck- - maybe try it in test-env later.

**Comment by ngaut** [comment]
You could try to use sysbench's point select benchmark with 256 threads.

**Comment by lysu** [comment]
update sysbench point-get with 256 result

before #11056 and #11055:

```
[ 10s ] thds: 256 tps: 164454.97 qps: 164454.97 (r/w/o: 164454.97/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 15s ] thds: 256 tps: 162979.98 qps: 162979.98 (r/w/o: 162979.98/0.00/0.00) lat (ms,95%): 2.91 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 256 tps: 162602.25 qps: 162602.25 (r/w/o: 162602.25/0.00/0.00) lat (ms,95%): 2.97 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 256 tps: 163254.63 qps: 163254.63 (r/w/o: 163254.63/0.00/0.00) lat (ms,95%): 2.91 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 256 tps: 163172.37 qps: 163172.37 (r/w/o: 163172.37/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 256 tps: 163340.89 qps: 163340.89 (r/w/o: 163340.89/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 256 tps: 164227.38 qps: 164227.38 (r/w/o: 164227.38/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 45s ] thds: 256 tps: 163061.99 qps: 163061.99 (r/w/o: 163061.99/0.00/0.00) lat (ms,95%): 2.91 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 256 tps: 163899.48 qps: 163899.28 (r/w/o: 163899.28/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 55s ] thds: 256 tps: 163417.61 qps: 163417.81 (r/w/o: 163417.81/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 256 tps: 164024.99 qps: 164024.79 (r/w/o: 164024.79/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 65s ] thds: 256 tps: 162233.85 qps: 162234.05 (r/w/o: 162234.05/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
```

after 

```
[ 15s ] thds: 256 tps: 166384.03 qps: 166384.03 (r/w/o: 166384.03/0.00/0.00) lat (ms,95%): 2.81 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 256 tps: 167907.39 qps: 167907.19 (r/w/o: 167907.19/0.00/0.00) lat (ms,95%): 2.76 err/s: 0.00 reconn/s: 0.00
[ 25s ] thds: 256 tps: 166532.81 qps: 166533.01 (r/w/o: 166533.01/0.00/0.00) lat (ms,95%): 2.81 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 256 tps: 166359.34 qps: 166359.34 (r/w/o: 166359.34/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 35s ] thds: 256 tps: 165876.74 qps: 165876.74 (r/w/o: 165876.74/0.00/0.00) lat (ms,95%): 2.81 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 256 tps: 166804.48 qps: 166804.48 (r/w/o: 166804.48/0.00/0.00) lat (ms,95%): 2.81 err/s: 0.00 reconn/s: 0.00
[ 45s ] thds: 256 tps: 167188.87 qps: 167188.87 (r/w/o: 167188.87/0.00/0.00) lat (ms,95%): 2.81 err/s: 0.00 reconn/s: 0.00
[ 50s ] thds: 256 tps: 161391.79 qps: 161391.79 (r/w/o: 161391.79/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 55s ] thds: 256 tps: 168173.47 qps: 168173.47 (r/w/o: 168173.47/0.00/0.00) lat (ms,95%): 2.76 err/s: 0.00 reconn/s: 0.00
[ 60s ] thds: 256 tps: 165497.13 qps: 165497.13 (r/w/o: 165497.13/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 65s ] thds: 256 tps: 166053.02 qps: 166053.02 (r/w/o: 166053.02/0.00/0.00) lat (ms,95%): 2.86 err/s: 0.00 reconn/s: 0.00
[ 70s ] thds: 256 tps: 167851.92 qps: 167851.92 (r/w/o: 167851.92/0.00/0.00) lat (ms,95%): 2.76 err/s: 0.00 reconn/s: 0.00
```

File: store/tikv/tikvrpc/tikvrpc.go [hunk 1]
```diff
@@ -233,90 +233,290 @@ func (req *Request) IsDebugReq() bool {
...
-	PessimisticRollback *kvrpcpb.PessimisticRollbackResponse
-
-	DebugGetRegionProperties *debugpb.GetRegionPropertiesResponse
-
-	Empty *tikvpb.BatchCommandsEmptyResponse
+	Resp interface{}
+}
+
+// Get returns GetResponse in response.
+func (resp *Response) Get() *kvrpcpb.GetResponse {
```
Comment (coocood) [inline store/tikv/tikvrpc/tikvrpc.go]: We can inline those methods as they are called only in one place.

**Comment by coocood** [comment]
LGTM

**Pull request merged**